### DISCUSSION TITLE
Morse code input for single button

### DISCUSSION THREAD

**Discussion by aaronze** [coelacant1/ProtoTracer#17] [Ideas]
While using single button input I've become frustrated with the lengthy process of navigating menus and the semi-random nature of switching expressions. Using a keypad would be a viable alternative but I think I can get similar levels of functionality out of the single button by adding a module to support morse code input and avoid having to lug around and operate a keypad. For example there's a few mappings that makes sense to me:
A - Angry
B - Blush
D - Doubt
F - Frown
H - Happy
M - Menu
O - Suprised
S - Sad

The actual implementation is quite simple for input streams that don't involve same-type only codes. For mixed inputs (dots and dashes) the algorithm is trivial:
1. Get input stream
2. Find shortest input X
3. Find longest input Y
4. Categorise inputs, durations closest to X are dots, durations closest to Y are dashes

For symbols with only dots or dashes, e.g. S (dot dot dot) you'd need to set some threshold for categorisation. Sadly I cannot find a way to make the dynamic duration algorithm work for all inputs, so the overall pseudocode would look something like:

1. Get input stream
2. Calculate average duration length
3. If all inputs are within some threshold of average duration: categorise all inputs as either dots or dashes based on some threshold
4. Else find shortest input X, longest input Y
5. Categorise inputs, durations closest to X are dots, durations closest to Y are dashes

I'm planning on implementing this at some point. Ideas and suggestions are welcome

**Comment by MeteorTheProto** [comment]
Only issue with this is that you need to have memorized Morse code. On my protogen I have an extra neopixel/rgb led on the inside so that I can change the color if it to match different expressions. Was thinking of extending this to the extra settings menus but I have those hard coded in for now.

**Reply by aaronze** [reply]
Sure, there is that caveat, I'm not suggesting it'd be a good idea to swap to using it as the default input method. I intend to add it as a module that can be enabled instead of single button support for those that want to use it.
